### DISCUSSION TITLE
Future firmware ideas?

### DISCUSSION THREAD

**Discussion by Nayphee** [frntc/RAD#22] [Ideas]
Just got my RAD to use with my RPi 3B+ not so long ago and having a good play.
VERY impressive piece of kit that lets the Pi really take over the C64.

Had a good play with changing the RAD menu music.
Only thing I wish existed was a WAV "jukebox" player in the firmware. I could create a directory of WAVs and easily switch between music files I've converted from the menu rather than having to overwrite the default music.wav file to hear the sample play.

"Bad Reception TV" format REU loading and booting the player works well. 
https://csdb.dk/release/?id=160140

Getting BRTV support built in would be a nice firmware features alongside nuvie.
I note that the BRTV sounds samples can hit higher quality sound rates (44100Hz), which is more than the RAD menu wav player.
Check out the "Personal Jesus 35KHz" BRTV REU image to hear what I mean. I even have a 44100Hz version which I downloaded from somewhere:
https://www.dropbox.com/sh/5qqxovv6y7osidc/AABeYf6PBwiL5Eqj7pnIImRZa?dl=0

Would it be possible to improve the WAV player in the menu to accept higher than 15600Hz bitrates and get better quality? 
Is there anything technically holding it back? Is it only because the RAD can stream to the SID at 22Khz tops? 
I did leave some converted wavs as 44100Hz and they went vveeeeeerrrryyyy ssssllllloooowwww! So bit rate seems to be more of a speed setting than a quality setting. But then, perhaps I'm comparing apples to oranges.

Finally if I had a dream, it would be for RAD to drive a version of RetroArch! Open the C64 up to "emulating" pretty much every other computer and arcade system!

**Comment by frntc** [comment]
The only reason for the playback rate was to keep the menu code simple (1 sample per rasterline, except for bad lines). In principle the RAD can write 1 sample per clock cycle. Maybe I'll allow a bit higher sample rates (but non-constant, otherwise the menu code becomes ugly). So far I didn't plan to include a WAV-player (btw the Sidekick64 does have one, including playback up to 48kHz stereo!).

BRTV: how many movie files are there?

RetroArch is a frontend requiring a full OS -- this won't work as the RAD runs baremetal (no OS). You can imagine that emulation of other platforms is possible (see RAD Doom), but someone has to port it to baremetal code ;-)

**Reply by Nayphee** [reply]
So not easy to port the Sidekick wav-player over?
There's a fair few BRTV videos. You can try them out at the links above as they work on the RAD already
Or this one: https://www.forum64.de/index.php?thread/69218-bad-reception-tv-br-tv/

Just on the RAD DOOM port? How was this done then? You must have compiled the code somehow....
Also, could the DOOM port be made to run the commercial WADs? Doom II and other compatible mods? Or was there some reason that it can only do DOOM shareware? Lack of memory for WADs or something?

**Reply by frntc** [reply]
It would be possible to port the Sidekick64-player, but writing it differently for the RAD would be really easy :-) but I have to admit that this is not very high on my list of things that I'd like to try out.

Thanks for the BRTV-link, maybe I'll add this. It actually only requires code that mounts the REU image plus autoloads the player (same as for NUVIEs).

Doom is compiled to bare metal code. That's DoomGeneric with minimal changes, so it can load what DoomGeneric can handle (and there are some parts in the code about WADs, but I didn't check). Memory is not an issue, I mean the RPi has >= 512MB ;-) My "goal" of RAD-Doom was to have a proof-of-concept of running some program with graphics and sound output on the C64, Doom was the instrument.

**Comment by drdoominator420** [comment]
is it possible to do some kind of youtube player?
or a nuvie player for ntsc systems?

**Reply by frntc** [reply]
well, in principle it would be easy to decode videos (all the code for playing is in RAD-Doom). However, one would need to download/buffer the videos (afaik that's not welcomed from that source...).

The Nuvie-format is made for PAL-machines, it's nothing I could do from the RAD side (RAD uses the standard Nuvie-player-PRG and just emulates a REU). There is a Koala-video-format that also runs on NTSC if that is of interest!?

**Reply by drdoominator420** [reply]
I wasnt aware of the koala video format. that should definitely be included along side the nuvie player.

**Comment by jajpohke** [comment]
here some crazy non scientist ideas from my rabbit's hat:
- a c64 doctorlike diagnostic tool able to write a more detailed log txt file that what we can do with dead test or with multiled diag charts;
- the zx spectrum baremetal emulator with ability to use audio jack to use speccy datassette :-) ;
- a built in Sonic The Edgedog distro (reu+game) so I dont have to plug other stuff to play it

**Reply by frntc** [reply]
Hi,

a diagnostic tool is something that I had on my list since my first project (the Sidekick64) -- unfortunately there were always other things I wanted to try before ;-)

The datasette will be write-only then, afaik the audio jack is output only :-) I haven't looked into something like this, but if there's a baremetal emulator, the RAD-Doom framework could be used to run it on the C64.

For Sonic: not sure about legal implications...

**Reply by jajpohke** [reply]
...about diagnostic and BMC64;
 In these day, as non pro(fit) hobbist, I'm trying to resurrect a broken c64c 250469, and RAD does what other carts dont. 

 I can successfull watch intro of RAD-DOOM with proper graphic and sound (with a working sid) even without the two 4464 bank of ram (or with bad ram in place) while a dead test would only give a blinking screen.  (I cant go any further than intro because I've desoldered CIAS and havent found time to socket and put back them in place).

I've search over the web if someone else was empyric like me to use a REU device to skip memory test of pcb and get some screen; I suppose so that a BMC64 on RAD device with CIAS could bring back many C64 without loosing a screw ;-)

note1:  saddly RAD regular reu+prg launcher realease, on that broken pcb,  starts with correct grayscaled borders and music but I've garbage garbage like in dead test (horizontal lines and debris) so finger crossed for superPla that I want to avoid the replacement :-D

**Comment by drdoominator420** [comment]
is a mp3 player possible?

**Reply by frntc** [reply]
A mp3-file could be decompressed and played like the menu-music (or with dual-SID, ... support). So yes, this is possible. I think mp3 is not covered by patents anymore...?

**Comment by xlar54** [comment]
I had wondered... If the raspberry pi (running Linux or bmc64) can emulate a 64 with Vice, would it be possible to allow the emulated Vice to send and recieve to the real hardware. There's your Scpu emulation. Just send the video back to the VIC-II.  Vice can run headless (no ui)

**Reply by StmLord** [reply]
running anything other OS than bare metal (eg Linux etc) will destroy the ability of the Pi to communicate with C64 because of timing delays, so no, I don't think that's possible. Except maybe if there is an emulator that run bare metal it can then be "constructed" to run via RAD-DOOM framework

**Reply by frntc** [reply]
I think it depends to which part of the real hardware you want to communicate. Reading keyboards, joysticks etc. is non time-critical, while loading from a real-floppy drive at the serial port is probably more tricky. (fortunately the expansion port is blocked once you put a RAD in, otherwise this would be even more critical). Ideally that emulator would run one cycle after another (fast enough). In any case, once you run an OS, you're pretty screwed as there are too many things going on (like thousands of interrupts per second which make reliable communication impossible (as StmLord said). In fact, my very first experiments prior to Sidekick64 used a RPi with Raspbian to in-place-emulate a SID, and communication was always unstable because of that.

**Reply by DirkS-71** [reply]
BMC64 is a bare Metal C64 Emulation on a Pi, so there is no OS below it. Using the RAD Doom Framework it might be possible to use part of the real C64, but for what reason?

**Comment by BizarreKING** [comment]
the RAD will be support another file format like crt o d64-d81?

**Reply by frntc** [reply]
The hardware itself is not capable of directly emulating all kinds of cartridges. This requires also emulating the CPU. While this code exists, I didn't find the time yet to make this ready for release.

D64: there's a huge difference of supporting the file format (e.g. via kernal vectors) or real IEC emulation. Again, the latter is something the hardware is not made for.

**Comment by NightriderC64** [comment]
Hi,
I was wondering about implementing some math coprocessor capabilities in RAD. In first step it could be integer operations like multiplication and division. Maybe some other functions supporting basic 2D/3D graphic transformation, mapping cartesian XY coordinates to bitmap...? If successful, the next step could be implementing some FPU functions. Also thinking about some Blitter capabilities, however REU's SWAP command does its job. So the goal could be REU and FPU mode in the same time? Making math operations on whole blocks of data, not only single byte/words? Maybe there are possibilities of extending this function...?
It could open some doors for new kind of demos/games. I know, purists would be disgusted, but who cares about them ;)

**Reply by frntc** [reply]
Hi, I was thinking about this since years (co-processor features could also be implemented into Sidekick64, SIDKick etc). My feeling is that there is little acceptance for such things as it would not emulate "hardware that existed back in the days". However, if someone wants to do some demo/game and solve the chicken and egg problem, the features would be easy to integrate.

**Reply by DirkS-71** [reply]
Well, concerning acceptance, I would see sth like a vektor multiplier pipeline like the original Geometrie Engine. You can always argue it is not your fault people didn't have an SGI beside their C64 in 1985 ;-) It would allow the C64 to render an huge amount of triangles. Elite might be upgradable with reasonable effort.
Seriously, I think providing a growing architecture of "simulated add ons" that could as well be fabricated physically like the turbo6510 you are working on, a v9958 secondary graphics card and or a superFX chip might buy some acceptance. The Pi3 should be able to handle that alltogether. 
I am currently trying to get into circle and setup a toolchain, so I hope I can support you in a few months

**Comment by tomnic79** [comment]
I'd really like to play Metal Dust on my real C64, so if DOOM can run why not that nice SuperCPU based game?

**Reply by frntc** [reply]
If you ask "could RAD do that?" the answer is most likely yes. But the SuperCPU emulation needs to be implemented first.

**Comment by tinspin** [comment]
NES PPU

**Reply by frntc** [reply]
I don't have a NES ;-)

**Comment by tinspin** [comment]
WiFi RR-Net with codenet?

Or maybe modem emulation?

Or maybe some new network thing?

What I lack is something that behaves like TCP (= no firewall problems) on the Pi but that behaves like a RxTx RS232 thing on the C64 side to not have overhead for multiplayer games, possibly MMO... How hard would that be?

For old school modems the Hayes protocol gave the DNS/IP part, but you could as well hack something custom together... from "hardcode the DNS/IP on the Pi" to some way to tell the Pi about a range from a hardcoded list somehow, to full blown send the domain/IP somehow... I don't know what would be simplest/possible?

Edit: I just realized I can do this myself if I only knew where/how the interactions worked. I'm going to start looking at the source. Basically I just need a fast and simple lane between the C64 and Zero where I can pass bytes back and forth. What is running on the Pi f.ex.

Aha https://github.com/rsta2/circle, hm mentions nothing about WiFi... How about running RaspiOS lite on the Pi is that out of the question? Ugh the boot time... Aha https://github.com/rsta2/circle/tree/master/addon/wlan/sample/hello_wlan

So now I just need to figure out the C64 side...

**Comment by tinspin** [comment]
This is not a firmware upgrade per se but I'm really looking for a ELI5 of the interfaces.

First what is the difference between the sidekick and the rad.

Which pins do what and how can you access them from the Pi/C64?

Even two simple basic+asm programs to POKE things over to the Pi and a simple C program to send things to the C64 would be nice!

Edit: also how do you even exit to basic, how do you configure the RAD to default to basic? Edit: You press X, how is one supposed to know that? Found it here https://github.com/frntc/RAD/discussions/2 "eXit"! Now I see the X is flashing... maybe E[X]IT would be better?

Hm, launching prgs does not work on reloaded v1.

Joystick in the menu would be nice?
